Additional docs/log for direct memory usage

jon-wei opened the pull request:
This PR adds a section to the performance tuning docs about how much direct memory Druid uses.

This PR also adds a log to IndexMerger when it allocates direct buffers for dictionary merging.

sascha-coenen commented:
Very useful documentation!

If its Ok to state this here, some formulations are not 100% clear to me still:
- "When a set of segments are merged during indexing". In which situations would segments be merged? Only in a Merge Task or any hadoop index task or only in re-indexing tasks? Is it correct to assume that "indexing" is something that only happens during the ingestion of data, so it is not happening on broker/historical, right? To my understanding, "indexing" is the creation of an index which would happen only during the ingestion of data, but then again, in the query phase, segments get scanned and segments contain indexes and they get merged and there are merge buffers, so I was wondering whether "segment merging during indexing" only refers to the index creation in the ingestion phase or also something that happens during query processing? I think its the first but I'm insecure...
- "fuzzy parameter meant to account for the decompression and dictionary merging buffers"
Prior to this phrase, the documentation mentions "merge buffers" but I got the feeling that "dictionary merge buffers" are yet something else. If so, it would be great to get a brief explanation of what they are and where they come into play (historicals? / brokers? etc) 
- "Indexing workers and historical"
I've been working with Druid for 18 months now but I'm still insecure about the use of the term "Indexing" as the documentation as a whole _explains_ terms like overlord/middlemanager/peons/realtime nodes but then _uses_ terms like "indexing service" or "indexing workers" and I don't know what exactly how these notions connect. What exactly is meant by "Indexing workers" in this case? Does it mean Peons/Middlemanagers? Other documentation also sometimes talks of "indexing service" in which case I'm not sure if "indexing service" is the short form of "Kafka Indexing service" or if "indexing-service" == overlord+middlemanagers+peons or if Indexing service == realtime nodes... I do feel that I understand all concepts and moving parts of Druid, it's just the terminology itself that throws me. A reoccuring pattern is that I read about similar terms and don't know if they are synonymous or not. 
- Below the newly added documentation, there is a section that says
"Larger values allow for more aggregations in a single pass over the data while smaller values can require more passes depending on the query that is being executed"
How can I tell whether Druid is needing several passes over the data? To my understanding, in order to achieve top performance, Druid should be doing only a single pass over the data. Therefore, I would like to make the processsing buffers sufficiently large such that for all query scenarios that can occur on my particular cluster, all query processing would be done in a single-pass fashion. So how can I tell whether that is actually happening? Lets say I have a testsuite covering all query scenarios over my particular dataset. How can I measure how many passes Druid ends up taking given my current memory settings? There is a segment scan time metric and if it is too high, that might indicate several passes, but thats just a fuzzy measure. Is there a way to read off how many passes were taken? Is this being logged or emitted as a metric?

leventov commented:
@kaijianding I think #4422 should solve your problem

jihoonson commented:
@kaijianding @leventov does groupBy v2 also use direct memory for both hash-based aggregation and result merge?

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -445,7 +445,9 @@ public int compare(Pair<Integer, PeekingIterator<String>> lhs, Pair<Integer, Pee
         if (useDirect) {
-          conversions[i] = ByteBuffer.allocateDirect(indexed.size() * Ints.BYTES).asIntBuffer();
+          int allocationSize = indexed.size() * Ints.BYTES;
+          log.info("Allocating dictionary merging direct buffer with size[%d]", allocationSize);
```
Recommend `%,d` to make this easier to read.

gianm replied:
There should also be a corresponding log message when the buffer is deallocated using `ByteBufferUtils.free(buf)`. And if it's never explicitly deallocated, then that is a problem! If that's the case, please either fix it or raise an issue for it.

jon-wei replied:
Hm, there isn't a corresponding deallocation, I'll do that in a follow on PR to this

gianm replied:
>  I'll do that in a follow on PR to this

thanks!

gianm commented on docs/content/operations/performance-faq.md:
```diff
@@ -20,6 +20,21 @@ We recommend 250mb * (processing.numThreads) for the heap.
 [Coordinator nodes](../design/coordinator.html) do not require off-heap memory and the heap is used for loading information about all segments to determine what segments need to be loaded, dropped, moved, or replicated.
 
+## How much direct memory does Druid use?
+Indexing workers and historical nodes use two kinds of direct memory buffers with configurable size: processing buffers and merge buffers.
```
Replace "Indexing workers and historical nodes" with "Any Druid node that handles queries" (broker is included too)

jon-wei replied:
Fixed this

gianm commented:
@kaijianding 

> As far as I know, only groupby v1 and topN query uses the reserved processing directly memory. The number of processing thread is limited due to the memory ls limited (like 16g or 32g per jvm).

groupBy v2 does use the processing buffers too, although if they are limited in size, performance shouldn't be hurt too much. So you can cut them down to 100MB or so and see how that works for you.

timeseries doesn't use them at all.

jon-wei commented:
@sascha-coenen Thanks for the comments!

>"When a set of segments are merged during indexing". In which situations would segments be merged? Only in a Merge Task or any hadoop index task or only in re-indexing tasks? Is it correct to assume that "indexing" is something that only happens during the ingestion of data, so it is not happening on broker/historical, right? To my understanding, "indexing" is the creation of an index which would happen only during the ingestion of data, but then again, in the query phase, segments get scanned and segments contain indexes and they get merged and there are merge buffers, so I was wondering whether "segment merging during indexing" only refers to the index creation in the ingestion phase or also something that happens during query processing? I think its the first but I'm insecure...

"indexing" used in this context is referring to "ingestion",  so merge tasks, hadoop index tasks, re-indexing tasks, kafka supervisor tasks, etc. can all involve segment merging. (From a code perspective, things that use IndexMerger).

I updated the doc changes in this PR to mention "ingestion" instead of "indexing" to be more clear.

> "fuzzy parameter meant to account for the decompression and dictionary merging buffers"
Prior to this phrase, the documentation mentions "merge buffers" but I got the feeling that "dictionary merge buffers" are yet something else. If so, it would be great to get a brief explanation of what they are and where they come into play (historicals? / brokers? etc)

The dictionary merge buffers are described in the second point under "Other sources of direct memory usage include:", I updated that with a point about how these buffers are independent of the other "merge buffers".

> "Indexing workers and historical"
I've been working with Druid for 18 months now but I'm still insecure about the use of the term "Indexing" as the documentation as a whole explains terms like overlord/middlemanager/peons/realtime nodes but then uses terms like "indexing service" or "indexing workers" and I don't know what exactly how these notions connect. What exactly is meant by "Indexing workers" in this case? Does it mean Peons/Middlemanagers? Other documentation also sometimes talks of "indexing service" in which case I'm not sure if "indexing service" is the short form of "Kafka Indexing service" or if "indexing-service" == overlord+middlemanagers+peons or if Indexing service == realtime nodes... I do feel that I understand all concepts and moving parts of Druid, it's just the terminology itself that throws me. A reoccuring pattern is that I read about similar terms and don't know if they are synonymous or not.

"Indexing workers" in this PR was referring to worker tasks that perform ingestion (peons), spawned by the overlord or middlemanager.
